### PR TITLE
Feature/Workspace API

### PR DESCRIPTION
**Pull request by quanpython**

### CODE REVIEW

File: studio/app/common/routers/workspace.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+- Workspace を作成する
+
+""",
+)
+def create_workspace(workspace: Workspace, db: Session = Depends(get_db)):
+    workspace = common_model.Workspace(**workspace.dict())
+    db.add(workspace)
+    db.commit()
+    db.refresh(workspace)
+    user = db.query(common_model.User).get(workspace.user_id)
```
Comment (itutu-tienday) [inline studio/app/common/routers/workspace.py:73]: - The process does not seem to go through. (500 error)
- workspace.user_id should be set to the ID of the currently logged-in user.

File: studio/app/common/routers/workspace.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    if not ws:
+        raise HTTPException(status_code=404)
+    data = workspace.dict(exclude_unset=True)
+    for key, value in data.items():
+        setattr(ws, key, value)
+    db.commit()
+    db.refresh(ws)
+    user = db.query(common_model.User).get(ws.user_id)
+    workspace.__dict__["user"] = user
+    return Workspace.from_orm(workspace)
```
Comment (itutu-tienday) [inline studio/app/common/routers/workspace.py:97]: Please check the delete flag.

File: studio/app/common/routers/workspace.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    dependencies=[Depends(get_current_user)],
+    description="""
+- Workspace を削除する
+""",
+)
+def delete_workspace(id: int, db: Session = Depends(get_db)):
+    ws = db.query(common_model.Workspace).get(id)
+    if not ws:
+        raise HTTPException(status_code=404)
+    db.delete(ws)
```
Comment (itutu-tienday) [inline studio/app/common/routers/workspace.py:112]: Deleting a workspace should be done with the delete flag On.

File: studio/app/common/routers/workspace.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    description="""
+- Workspace を更新する
+""",
+)
+def update_workspace(id: int, workspace: Workspace, db: Session = Depends(get_db)):
+    ws = db.query(common_model.Workspace).get(id)
+    if not ws:
+        raise HTTPException(status_code=404)
+    data = workspace.dict(exclude_unset=True)
+    for key, value in data.items():
```
Comment (itutu-tienday) [inline studio/app/common/routers/workspace.py:91]: Are there items (primary key:id) that do not need to be updated that are being updated?

File: studio/app/common/routers/workspace.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    response_model=bool,
+    dependencies=[Depends(get_current_user)],
+    description="""
+- Workspace の共有状態を更新する（総入れ替え）
+""",
+)
+def update_workspace_share_status(
+    id: int, data: WorkspaceSharePostStatus, db: Session = Depends(get_db)
+):
+    workspace = db.query(common_model.Workspace).get(id)
```
Comment (itutu-tienday) [inline studio/app/common/routers/workspace.py:178]: Please check the delete flag.

File: studio/app/common/routers/workspace.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+@router.get(
+    "/workspace/share/{id}/status",
+    response_model=WorkspaceShareStatus,
+    dependencies=[Depends(get_current_user)],
+    description="""
+- Workspace の共有状態を取得する
+""",
+)
+def get_workspace_share_status(id: int, db: Session = Depends(get_db)):
+    workspace = db.query(common_model.Workspace).get(id)
```
Comment (itutu-tienday) [inline studio/app/common/routers/workspace.py:152]: Please check the delete flag.

File: studio/app/common/routers/workspace.py [hunk 7]
```diff
@@ -143,13 +185,24 @@ def import_workspace(
...
+        .filter(
+            common_model.Workspace.id == id,
+            common_model.Workspace.user_id == current_user.id,
+            common_model.Workspace.deleted.is_(False),
+        )
+        .first()
+    )
     if not workspace:
         raise HTTPException(status_code=404)
     users = (
```
Comment (itutu-tienday) [inline studio/app/common/routers/workspace.py:208]: Please check users.active.

**Review by itutu-tienday — CHANGES_REQUESTED** [review]
@quanpython 
I have added the remarks. Please confirm.

File: studio/__main_unit__.py [hunk 8]
```diff
@@ -35,12 +36,13 @@
 app.include_router(run.router)
 app.include_router(users_admin.router)
 app.include_router(users_me.router)
-app.include_router(expdb.router, tags=["Experiment Database"])
+app.include_router(workspace.router, tags=["Workspace"])
```
Comment (ReiHashimoto) [inline studio/__main_unit__.py:39]: Just coding style, please set tags in `router = APIRouter(...)`declaration like other routers.

File: studio/app/common/core/auth/auth_dependencies.py [hunk 9]
```diff
@@ -24,8 +28,14 @@ async def get_current_user(
...
-            authed_user: User = await get_user(user["uid"])
-            return authed_user
+            authed_user, role_id = (
+                db.query(UserModel, UserRoleModel.role_id)
+                .outerjoin(UserRoleModel, UserRoleModel.user_id == UserModel.id)
+                .filter(UserModel.uid == user["uid"], UserModel.active.is_(True))
+                .first()
+            )
+            authed_user.__dict__["role_id"] = role_id
+            return User.from_orm(authed_user)
```
Comment (ReiHashimoto) [inline studio/app/common/core/auth/auth_dependencies.py:38]: Could you refactor to separate this process into another function?
Almost same logic are used in both using firebase_token and using ex_token.

File: studio/app/common/routers/workspace.py (hunk 7 above)
Comment (quanpython) [reply studio/app/common/routers/workspace.py:208]: I already check active user in `get_current_user` dependence. Is it reasonable?

File: studio/config/docker/Dockerfile [hunk 10]
```diff
@@ -24,4 +24,6 @@ COPY main.py pyproject.toml alembic.ini ./
 RUN pip3 install --no-cache-dir --upgrade pip && \
     pip3 install --no-cache-dir -e .
 
+RUN alembic upgrade head
+
```
Comment (ReiHashimoto) [inline studio/config/docker/Dockerfile:28]: This code was removed in #52  and this PR seems to include commits in #52 .
Could you rebase to the latest develop-main?

**Review by ReiHashimoto — CHANGES_REQUESTED** [review]

File: studio/app/common/routers/workspace.py (hunk 7 above)
Comment (itutu-tienday) [reply studio/app/common/routers/workspace.py:208]: @quanpython 
Please check the active flag of the table (Common_model.User) that joins with WorkspacesShareUser, not the active flag of the logged in user (yourself).

File: studio/app/common/core/auth/auth_dependencies.py (hunk 9 above)
Comment (ReiHashimoto) [reply studio/app/common/core/auth/auth_dependencies.py]: @quanpython 
smart refactor, thank you.

**Review by ReiHashimoto — APPROVED** [review]
LGTM

**Review by itutu-tienday — APPROVED** [review]

**Comment by itutu-tienday** [comment]
@quanpython 
We have identified a series of corrections. Thank you very much.
We will proceed to the merge to develop-main once the Conflict has been resolved.

**Pull request merged**